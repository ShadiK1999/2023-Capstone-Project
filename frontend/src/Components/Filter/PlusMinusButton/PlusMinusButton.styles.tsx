import { style } from 'typestyle';

export const Container = style({
  display: 'flex',
  height: '53px',
  width: '100%',
  alignItems: 'center',
  border: '1px solid black',
  borderRadius: '6px',
});

export const ButtonLeft = style({
  order: 1,
  alignSelf: 'stretch',
  width: '50px',
  display: 'flex',
  alignItems: 'center',
  justifyContent: 'center',
  borderRight: '1px solid black',
  $nest: {
    '&:hover': {
      backgroundColor: '#BFCCB5',
      borderTopLeftRadius: '6px',
      borderBottomLeftRadius: '6px',
    },
  },
});

export const ButtonContent = style({
  order: 2,
  flexGrow: 1,
  whiteSpace: 'nowrap',
  textAlign: 'center',
  padding: '0 15px',
});

export const ButtonRight = style({
  order: 3,
  alignSelf: 'stretch',
  width: '50px',
  display: 'flex',
  alignItems: 'center',
  justifyContent: 'center',
  borderLeft: '1px solid black',
  $nest: {
    '&:hover': {
      backgroundColor: '#BFCCB5',
      borderTopRightRadius: '6px',
      borderBottomRightRadius: '6px',
    },
  },
});
