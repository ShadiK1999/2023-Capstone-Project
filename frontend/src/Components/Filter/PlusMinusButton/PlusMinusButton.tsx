import * as style from './PlusMinusButton.styles';
import filterIcon from './Assets/Filter.svg';
import plusIcon from './Assets/Plus.svg';
import minusIcon from './Assets/Minus.svg';

const Icons = {
  FILTER: filterIcon,
  PLUS: plusIcon,
  MINUS: minusIcon,
};

type IconName = keyof typeof Icons;

type ButtonAttributes = {
  icon?: IconName;
  onClick: () => void;
};

type ButtonProps = {
  leftButton?: ButtonAttributes;
  rightButton?: ButtonAttributes;
  children: React.ReactNode;
};

const defaultLeftButton: ButtonAttributes = {
  icon: 'MINUS',
  onClick: () => {},
};

const defaultRightButton: ButtonAttributes = {
  icon: 'PLUS',
  onClick: () => {},
};

const PlusMinusButton: React.FC<ButtonProps> = ({
  leftButton,
  rightButton,
  children,
}: ButtonProps) => {
  const { icon: leftIcon, onClick: leftOnClick } = Object.assign({}, defaultLeftButton, leftButton);
  const { icon: rightIcon, onClick: rightOnClick } = Object.assign(
    {},
    defaultRightButton,
    rightButton,
  );

  return (
    <div className={style.Container}>
      <div className={style.ButtonLeft} onClick={leftOnClick}>
        <img src={leftIcon && Icons[leftIcon]} alt={leftIcon && leftIcon} />
      </div>
      <div className={style.ButtonContent}>{children}</div>
      <div className={style.ButtonRight} onClick={rightOnClick}>
        <img src={rightIcon && Icons[rightIcon]} alt={rightIcon && rightIcon} />
      </div>
    </div>
  );
};

export { PlusMinusButton };
