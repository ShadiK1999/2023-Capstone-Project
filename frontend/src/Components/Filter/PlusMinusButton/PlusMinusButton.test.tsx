import { render } from '@testing-library/react';
import { PlusMinusButton } from './PlusMinusButton';
import { describe, it, expect, vi, afterEach, beforeEach } from 'vitest';

describe('Filter', () => {
  beforeEach(() => {
    vi.useFakeTimers();
    vi.setSystemTime(new Date(2023, 8, 4));
  });

  afterEach(() => {
    vi.useRealTimers();
  });

  it('Matches DOM Snapshot', () => {
    const tree = render(
      <PlusMinusButton
        leftButton={{ onClick: () => void 0 }}
        rightButton={{ onClick: () => void 0 }}
      >
        Hello World
      </PlusMinusButton>,
    );
    expect(tree).toMatchSnapshot();
  });

  it('renders default icons if none provided', () => {
    const { getByAltText } = render(
      <PlusMinusButton
        leftButton={{ onClick: () => void 0 }}
        rightButton={{ onClick: () => void 0 }}
      >
        Hello World
      </PlusMinusButton>,
    );
    expect(getByAltText('MINUS')).toBeInTheDocument();
    expect(getByAltText('PLUS')).toBeInTheDocument();
  });

  it('renders custom icons if provided', () => {
    const { getByAltText } = render(
      <PlusMinusButton
        leftButton={{ icon: 'FILTER', onClick: () => void 0 }}
        rightButton={{ onClick: () => void 0 }}
      >
        Hello World
      </PlusMinusButton>,
    );
    expect(getByAltText('FILTER')).toBeInTheDocument();
  });

  it('Fires onClick event', () => {
    const onClick = vi.fn();
    const { getByAltText } = render(
      <PlusMinusButton leftButton={{ onClick }} rightButton={{ onClick: () => void 0 }}>
        Hello World
      </PlusMinusButton>,
    );
    const button = getByAltText('MINUS');
    button.click();
    expect(onClick).toHaveBeenCalled();
  });
});
