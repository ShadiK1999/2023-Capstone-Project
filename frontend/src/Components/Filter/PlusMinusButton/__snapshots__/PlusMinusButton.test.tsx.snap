// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Filter > Matches DOM Snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="f723edx"
      >
        <div
          class="f2ey5g1"
        >
          <img
            alt="MINUS"
            src="/src/Components/Filter/PlusMinusButton/Assets/Minus.svg"
          />
        </div>
        <div
          class="f1sz7xol"
        >
          Hello World
        </div>
        <div
          class="f1fozamg"
        >
          <img
            alt="PLUS"
            src="/src/Components/Filter/PlusMinusButton/Assets/Plus.svg"
          />
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="f723edx"
    >
      <div
        class="f2ey5g1"
      >
        <img
          alt="MINUS"
          src="/src/Components/Filter/PlusMinusButton/Assets/Minus.svg"
        />
      </div>
      <div
        class="f1sz7xol"
      >
        Hello World
      </div>
      <div
        class="f1fozamg"
      >
        <img
          alt="PLUS"
          src="/src/Components/Filter/PlusMinusButton/Assets/Plus.svg"
        />
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
